### PR TITLE
Heavy Ions MVA track selection + pt dependent pixel track selection

### PR DESCRIPTION
**Pull request by istaslis**
Addition of MVA track selection for heavy ions.
As agreed with @VinInn this was done by implementing a separate version of the multi-track selector in RecoHI/HiTracking.  
We should remerge with the pp version in the future when the pp version allows a more flexible variable selection (which is a planned future development).
The MVA track selection is _disabled_ by default, as the calibrations are not yet available in the GT.  
The calibrations are uploaded and the GT has been requested, so we should be enable this feature shortly, which we propose to do by a separate PR (this can be discussed at the ORP tomorrow).

This PR also included the pT dependent pixel track selection cuts implemented in the now closed PR #9017, as agreed with @VinInn

### CODE REVIEW

**Comment by appeltel** [comment]
I don't believe that the test error is related to the PR - I saw similar
DAS Errors today performing "runTheMatrix.py -s" on unrelated branches, and
these wen't away after retesting.

-Eric

**Comment by mandrenguyen** [comment]
@cmsbuild please test
Let's see if I have the rights to trigger a retest

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@VinInn @rovere 
What is the time scale to get MultiTrackSelector refactored to handle the features needed for HI?
HI tracking code is still tracking POG code; it would be nice to keep things coherent.

Since that would be a technical change,  we likely can have it in pre6 (~3 weeks), if it can be ready by then. Otherwise, timescale of July-August for 76X is the next target.

**Comment by slava77** [comment]
@istaslis please update the subject of the PR to something more readable (there's an "Edit" button at the top of the PR interface page)

**Comment by VinInn** [comment]
@mhwalker will soon publish a new PR that handle the new pp MVA.
Once that is in, We will refactor the class to more flexibly handle different type of variable set.
This will be done using C++ "plugin", so a specific HI variable set will still need to be coded

File: RecoHI/HiTracking/plugins/HIMultiTrackSelector.cc [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+    mvavalues[minlost] = std::min(lostIn,lostOut);
+    mvavalues[lostmidfrac] = trk.numberOfLostHits() / (trk.numberOfValidHits() + trk.numberOfLostHits());
+
+    mvavalues[etaerror] = trk.etaError();
+
+    float reldz = 0;
+    float reldxy = 0;
+    if (checkvertex) {
+      int vtxind = 0; // only first vertex is taken into account for the speed purposes
+      float dxy = trk.dxy(vertices[vtxind].position()), dxyE =  sqrt(trk.dxyError()*trk.dxyError()+vertices[vtxind].xError()*vertices[vtxind].yError());
```
Comment (slava77) [inline RecoHI/HiTracking/plugins/HIMultiTrackSelector.cc]: dxyE is mildly incorrect (OK for symmetric vertices): the vertex error component should be added in the direction orthogonal to pt

**Comment by slava77** [comment]
... I'm checking the PR

I looked at the code, comparing with the original MultiTrackSelector.cc and didn't find anything particularly out of order.
Now to check runtime.

From jenkins shorter run in data (only 140.53 is in the matrix):
- there is an increase in low-pt tracks
all_oldvsnew_runhi2011wf140p53c_log10recotracks_higeneraltracks__rereco_obj_p
- initial step yield decreases, while lowPtTriplet is up by ~30% and detachedTriplet up by about 50%
all_oldvsnew_runhi2011wf140p53c_log10recotracks_higeneraltracks__rereco_obj_pt
all_oldvsnew_runhi2011wf140p53c_recotracks_higeneraltracks__rereco_obj_algo
- major increase (x4 in 8-9 hit tracks)
all_oldvsnew_runhi2011wf140p53c_recotracks_higeneraltracks__rereco_obj_found

Since this is data, no idea if it's fakes or real tracks

**Comment by mandrenguyen** [comment]
@slava77 
In the initial step of HI tracking all reconstructed tracks were kept (prior to this PR), for backwards compatibility.
As we added further iterations we decided to use the 'loose' bit to define what is kept in the output collection (keepAllTracks = False).
With the new MVA track selection we now apply a uniform loose selection to all steps, which removes tracks which are considered unrecoverable in central events for any reasonable track selection.  The effect on the efficiency was checked to be on the order of 1%.
The net effect on the total output size is rather small because, as you saw, there are less tracks in the initial step, and more from the subsequent steps. 
The tracks which are added will actually increase the efficiency by a fair amount once the MVA selection is activated.

**Comment by slava77** [comment]
based on wflow 140.0 (minbias PbPb events)

high-purity selections show increase in efficiency:
wf140_hp_algo
wf140_hp_eff_vs_pt
fake rate about doubles
wf140_hp_fake_vs_eta
wf140_hp_fake_vs_pt

for loose tracks: the efficiency and the fake rate is about the same, for a noticeable change in the algo values
wf140_reco_algo
cf to all hiGeneralTracks
all_sign550histatvsorig-9000histat_hydjetqminbiaswf140p0c_recotracks_higeneraltracks__reco_obj_algo

**Comment by slava77** [comment]
Technical performance numbers base on 100 events in 140.53:
- event sizes increase starting from more tracks which contribute to objects downstream
    - 9% more hiGeneralTracks
    - 5.4% more muons
    - 6% more particleFlow candidated
    - 6 to 12% more akVs PF jets
    - 2.3% increase in RECO size
    - 2.5% increase in AOD size
- CPU consumption is up globally by about 14%, driven primarily by increase in  particleFlowEGamma and voronoiBackgroundPF. In the order of execution, the increase is starting mainly from the track merging step (smaller increases/decreases appear in the following (not show, the iterative steps times are down by ~10% in pixel pair and low-pt triplet steps without noticeable increases in other steps).
```
        hiGeneralTracks          559.082 ms/ev -> 640.416 ms/ev
        pfTrack          46.43 ms/ev -> 64.173 ms/ev
        pfTrackElec      916.089 ms/ev -> 1111.54 ms/ev
        particleFlowBlock        9107.72 ms/ev -> 9915.89 ms/ev
        particleFlowEGamma       10588.7 ms/ev -> 14557 ms/ev
        particleFlowTmp          1615.37 ms/ev -> 2163.43 ms/ev
        voronoiBackgroundPF      8097.48 ms/ev -> 12685.9 ms/ev
 Total times (all reco step modules) : 71511.4 ms/ev -> 81320.1 ms/ev     delta: 9808.68
```

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/9032	d916c06
- tested locally in CMSSW_7_5_0_pre4 /test area sign550/, cherry-picked
- changes appear only in HI workflows as expected 
- changes in HI workflows observed in 140.53 and 140.0 appear to be as expected (although the increase in CPU is probably putting a more significant squeeze on HI program)

**Comment by davidlange6** [comment]
+1

**Pull request merged**